Shoko.Server/Plugin/Loader.cs

```diff
@@ -18,36 +20,42 @@ public class Loader : ISettingsProvider
         private readonly IList<Type> _pluginTypes = new List<Type>();
         private static readonly ILogger Logger = LogManager.GetCurrentClassLogger();
         
-        internal void Load(IServiceCollection serviceCollection)
+        internal void Load(IServiceCollection serviceCollection, IConfiguration config)
         {
             var assemblies = new List<Assembly>();
             var assembly = Assembly.GetExecutingAssembly();
             var uri = new UriBuilder(assembly.GetName().CodeBase);
             var dirname = Path.GetDirectoryName(Uri.UnescapeDataString(uri.Path));
             // if (dirname == null) return;
             assemblies.Add(Assembly.GetCallingAssembly()); //add this to dynamically load as well.
-            
+            var pluginConfig = config.GetSection("Plugins");
+            var pluginSettings = pluginConfig.Get<PluginSettings>();
+
             //Load plugins from the user config dir too.
             var userPluginDir = Path.Combine(ServerSettings.ApplicationPath, "plugins");
-            var userPlugins = Directory.Exists(userPluginDir) ? Directory.GetFiles(userPluginDir) : new string[0];
+            var userPlugins = Directory.Exists(userPluginDir) ? Directory.GetFiles(userPluginDir, "*.dll") : new string[0];
             
             foreach (var dll in Directory.GetFiles(dirname, "plugins/*.dll", SearchOption.AllDirectories).Concat(userPlugins))
             {
                 try
                 {
                     string name = Path.GetFileNameWithoutExtension(dll);
-                    if (ServerSettings.Instance.Plugins.EnabledPlugins.ContainsKey(name) &&
-                        !ServerSettings.Instance.Plugins.EnabledPlugins[name])
+                    if (pluginSettings.EnabledPlugins.ContainsKey(name) && pluginSettings.EnabledPlugins[name] == false)
```

Comment:
Ultimately, the discovery will have to be moved outside of the enabled state so you follow the current flow:

Discover DLLs.
Do the container injection for plugins.
Finish DI container
Instantiate based on enabled settings.
Inject in the settings

Reply:
That would probably be the best flow, since editing settings before the service provider is built is probably unconventional.